### PR TITLE
Shorten def_span of closures to just their header

### PR DESCRIPTION
**Pull request by cjgillot**
Continuation of https://github.com/rust-lang/rust/pull/93967.

### CODE REVIEW

File: src/test/ui/associated-types/impl-wf-cycle-1.stderr [hunk 1]
```diff
@@ -1,14 +1,8 @@
...
   --> $DIR/impl-wf-cycle-1.rs:15:1
    |
-LL | / impl<T: Grault> Grault for (T,)
-LL | | where
-LL | |     Self::A: Baz,
-LL | |     Self::B: Fiz,
-...  |
-LL | |
-LL | | }
-   | |_^
```
Comment (estebank) [inline src/test/ui/associated-types/impl-wf-cycle-1.stderr:11]: Interestingly, we should probably point at the first `where` clause in this error as the primary span, but that's out of scope.

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 2]
```diff
@@ -947,28 +947,106 @@ impl<'hir> Map<'hir> {
...
+                span: outer_span,
+                ..
+            })
+            | Node::TraitItem(TraitItem {
+                kind: TraitItemKind::Type(bounds, _),
+                generics,
+                span: outer_span,
+                ..
+            }) => {
+                let end = if let Some(b) = bounds.last() { b.span() } else { generics.span };
```
Comment (estebank) [inline compiler/rustc_middle/src/hir/map/mod.rs:1021]: Whould we also account for the `where` clause here when appropriate? (I think that `type` is the only case where you wouldn't want to due to syntactical ordering.)

File: src/test/ui/borrowck/mutability-errors.stderr [hunk 3]
```diff
@@ -119,146 +119,96 @@ LL |     &mut (*f()).0;
...
-   |  _____------_-
-   | |     |
-   | |     expects `Fn` instead of `FnMut`
-LL | |         x = (1,);
-LL | |         x.0 = 1;
-LL | |         &mut x;
-LL | |         &mut x.0;
-   | |         ^^^^^^^^ cannot borrow as mutable
-LL | |     });
-   | |_____- in this closure
```
Comment (estebank) [inline src/test/ui/borrowck/mutability-errors.stderr:261]: This looks like a regression to me. We are not pointing at the closure header. We probably have some extra logic to see if the closure is in a single line that is imperfect and now doesn't trigger even when it should.

**Review by estebank — APPROVED** [review]
r=me after rebasing and optionally addressing the regression in E0596 (or filing a follow up ticket). Make it rollup=never.

File: src/tools/clippy/tests/ui/integer_arithmetic.stderr [hunk 4]
```diff
@@ -123,6 +123,18 @@ error: integer arithmetic detected
 LL |     i >>= 2;
    |     ^^^^^^^
 
+error: integer arithmetic detected
+  --> $DIR/integer_arithmetic.rs:68:28
+   |
+LL |         const ASSOC: i32 = 1 + 1;
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/integer_arithmetic.stderr:129]: Why does this change here and in the other similar PR. This has something to do with associated constants.

**Review by flip1995 — APPROVED** [review]
Clippy changes LGTM now.

**Comment by estebank** [comment]
r=me after rebasing & reblessing

**Comment by cjgillot** [comment]
The regression is mostly in `tt-muncher` case.  From the detailed view, it would be in `expand_crate`.  I don't see how a change in error-reporting span can cause such a performance change.

**Pull request merged**